Wrote a tutorial for a CREATE2 factory (#1024)
foundry-rs/book: src/tutorials/create2-tutorial.md

## Deterministic deployment using CREATE2

### Introduction

> Enshrined into the EVM as part of the [Constantinople fork](https://ethereum.org/en/history/#constantinople) of 2019, `CREATE2` is an opcode that started it's journey as [EIP-1014](https://eips.ethereum.org/EIPS/eip-1014).
> It is often, rather fancifully IMO, referred to as the opcode that allows you to "*interact with addresses that haven't been created yet*".
> What this basically means is that `CREATE2` allows you to deploy smart contracts to deterministic addresses, based on parameters that you control.
> This stands in stark contrast to the `CREATE` opcode, where the nonce of the deploying address serves as one of the parameters that determines the address of the deployed contract. With `CREATE2`, you can use the same address to deploy smart contracts to the same address across multiple networks, even if the address has varying nonces.

Reviewer: Just rephrasing a bit to more closely match the style/tone of other docs pages :)

```suggestion
The `CREATE2` opcode was added to the EVM as part of the [Constantinople fork](https://ethereum.org/en/history/#constantinople) in 2019, and is specified by [EIP-1014](https://eips.ethereum.org/EIPS/eip-1014).
`CREATE2` allows you to deploy smart contracts to deterministic addresses, based on parameters controlled by the deployer. 
As a result, it's often mentioned as enabling "counterfactual" deployments, where you can interact with an addresses that haven't been created yet because CREATE2 guarantees known code can be placed at that address.
This is in contrast to the `CREATE` opcode, where the address of the deployed contract is a function of the deployer's nonce.
 With `CREATE2`, you can use the same deployer account to deploy contracts to the same address across multiple networks, even if the address has varying nonces.
```

Author: Is it possible to keep just the first sentence the way it was originally?

"_Enshrined into the EVM as part of the [Constantinople fork](https://ethereum.org/en/history/#constantinople) of 2019, `CREATE2` is an opcode that started it's journey as [EIP-1014](https://eips.ethereum.org/EIPS/eip-1014)._" 

I'll edit the rest of the paragraph according to your suggestions. I gave the initial sentences a loooot of thought to make an engaging hook to the article, hence the poetic language. 
Would love to just keep the first sentence if possible.

In this tutorial, we will:

> 1. Look at a `CREATE2` factory implementation.

Reviewer: It might be useful to add a note somewhere, along the lines of: "This guide is intended to help understand CREATE2. In most use cases, you won't need to write and use your own deployer, and can use an existing deterministic deployer. In forge scripts, using `new MyContract{salt: salt}()` will use the deterministic deployer at [`0x4e59b44847b379578588920ca78fbf26c0b4956c`](https://github.com/Arachnid/deterministic-deployment-proxy)."

Author: Will do. To be fair, Arachnid needs to update the README of the repo to reflect the fact that the Yul CREATE2 deployer has actually been deployed to many more networks.

2. Deploy the factory using the traditional deployment methods.

> 3. Use this deployed factory to in turn deploy the same `CREATE2` factory at a deterministic address.

Reviewer: I'd use a simple `Counter` contract here instead, such as the one you get with `forge init`. I think readers might find it confusing if the guide has a contract deploy an identical copy of itself

4. Simulate this set of events by writing a fork test in Foundry.

### Prerequisites

> 1. Some familiarity with Solidity and Foundry is required, and some familiarity with the in-line assembly is recommended.
> Refer [official Solidity docs](https://docs.soliditylang.org/en/latest/assembly.html) for a primer on inline assembly.

Reviewer: ```suggestion
1. Some familiarity with Solidity and Foundry is required, and some familiarity with the inline assembly is recommended.
Refer to the [official Solidity docs](https://docs.soliditylang.org/en/latest/assembly.html) for a primer on inline assembly.
```

> 2. Make sure you have Foundry installed in your system. [Refer Installation](../getting-started/installation.md).
> 3. Initialize a local Foundry project. [Refer Creating a New Project](../projects/creating-a-new-project.md).

Reviewer: ```suggestion
2. Make sure you have Foundry [installed](../getting-started/installation.md) on your system.
3. [Initialize](../projects/creating-a-new-project.md) a new Foundry project.
```

### CREATE2 Factory

Create a file named `Create2.sol` Inside the `src` directory.
Initialize a contract named `Create2` like this:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.20;

contract Create2 {

    error Create2InsufficientBalance(uint256 received, uint256 minimumNeeded);

    error Create2EmptyBytecode();

    error Create2FailedDeployment();
}
```

> These self-explanatory errors are meant to enforce some sanity checks on the factory deployment, and revert the whole transaction when triggered.

Reviewer: I'd suggest giving a brief explanation of each, since e.g. "failed deployment is" pretty vague—why did deployment fail?

Author: Done

> Next, initialize a function named deploy like this:

Reviewer: ```suggestion
Next, create a function named `deploy`:
```

```solidity

>  function deploy(uint256 amount, bytes32 salt, bytes memory bytecode) external payable returns (address addr) {

Reviewer: I'd suggest deleting the `amount` parameter and just use `msg.value` as the amount of ETH to send with the deploy—this is a common approach for create2 deployers. This also removes the need for the `Create2InsufficientBalance` error. I'd also rename `bytecode` to `initializationCode` or `creationCode` to make it clear we need that (as opposed to the runtime bytecode)

Author: Replacing `bytecode` with `creationCode` to more accurately reflect the sentiment here.
Also using  `creationCode` instead of `initializationCode` because Foundry also usesuses that term:
`type(Counter).creationCode`

 
 }
```

> This function will take in 3 params, and return the address of the newly deployed contract:
> 
> 1. The random `salt` used to calculate the final address.
> 2. The bytecode of the contract that we want to deploy.
> 3. The amount of ETH that we want to send to the new address as part of the deployment.

Reviewer: ```suggestion
This function takes 3 inputs:

1. The `salt` used to calculate the final address.
2. The creation code of the contract that we want to deploy.
3. The `msg.value` is an implicit input and is the amount of ETH sent to the new address as part of the deployment.

The address of the newly deployed contract is the returned after a successful deploy.
```

> ℹ️ **Note**
>
> You can send ETH to a contract that is being deployed using CREATE2, but only if it has a payable constructor.
> If you try to send ETH to it without a payable constructor, the transaction will revert.

Add these two revert statements at the top of the deploy function. We want our function to reject the deploy request if the following conditions are not met:

```solidity
        if (msg.value < amount) {
            revert Create2InsufficientBalance(msg.value, amount);
        }

        if (bytecode.length == 0) {
            revert Create2EmptyBytecode();

>         }

Reviewer: Is this check necessary? I _think_ the EVM may revert if you pass empty bytecode but would need to test it to confirm

Author: Lemme test it ser :)

Author: @mds1 check this out.

1. I removed the `Create2EmptyBytecode()` error.
2. Initialized an empty `creationCode` variable in the test file.
3. **The deployment still passes.** Lemme know if I am doing anything wrong.
![image](https://github.com/foundry-rs/book/assets/56264430/62a2e73a-0d38-452e-9eec-25858de9115b)
![image](https://github.com/foundry-rs/book/assets/56264430/3330c805-c87f-49de-bb05-bf59ee80c433)

![image](https://github.com/foundry-rs/book/assets/56264430/f3dc1184-1a80-4ee2-985d-5dc72c93e730)

Author: You can see in last SS that I am passing an empty bytecode to deploy function

```

> Next, we will call the `CREATE2` opcode using in-line assembly, which can be done using the `assembly` keyword:

Reviewer: ```suggestion
Next, we will call the `CREATE2` opcode using inline assembly, which can be done using the `assembly` keyword:
```

```solidity
        assembly {

>             addr := create2(amount, add(bytecode, 0x20), mload(bytecode), salt)

Reviewer: ```suggestion
            addr := create2(callvalue(), add(creationCode, 0x20), mload(creationCode), salt)
```

Author: I had no idea `callvalue()` was a thing. Thanks. Because you can't use `msg.sender` directly in inline assembly

        }
```

> To call [the CREATE2 opcode from in-line assembly](https://docs.soliditylang.org/en/latest/yul.html#evm-dialect), we need to pass in 4 parameters:

Reviewer: I'd move this block before the assembly snippet immediately above, i.e. explain what we'll do, then show how to do it
```suggestion
To call [the CREATE2 opcode from inline assembly](https://docs.soliditylang.org/en/latest/yul.html#evm-dialect), we need to pass in 4 parameters:
```

Author: Done

1. The amount of ETH that we want to send to the new address as part of the deployment.

> 2. The second and third parameters refer to the range of memory our bytecode is located in. `add(bytecode, 0x20)` takes in a reference to the location of the `bytes` variable bytecode in memory, and skips 32 bytes(0x20 in hex) to point to the actual bytecode.

Reviewer: ```suggestion
2. The second and third parameters refer to the range of memory our bytecode is located in. `add(bytecode, 0x20)` takes in a reference to the location of the `bytes` variable bytecode in memory, and skips 32 bytes (0x20 in hex) to point to the actual bytecode.
```

`mload(bytecode)` loads the length of the bytecode into memory.

> ℹ️ **Note**
>

> > The `bytes` type in Solidity is a dynamically sized byte array, and is stored in memory as a 32 byte value, where the first 32 bytes
> > represent the length of the array, and the rest of the bytes represent the actual array. Therefore, when we pass in a reference to
> > `bytes` variable, we need to skip the first 32 bytes to point to the actual array.
> > Read more about the `add` and `mload` opcodes at [evm.codes](https://www.evm.codes/?fork=shanghai).

Reviewer: ```suggestion
> The `bytes` type in Solidity is a dynamically sized byte array, where the first 32 bytes of memory
> represent the length of the array, and the remaining bytes represent the actual data. Therefore, when we pass in a reference to
> `bytes` variable, we need to skip the first 32 bytes to point to the actual data.
> Read more about the `add` and `mload` opcodes at [evm.codes](https://www.evm.codes/?fork=shanghai).
```

> Finally, we will revert the whole transaction if the deployment fails for any reason, in which case the CREATE2 opcode will return a 0 address:

Reviewer: Consider mentioning reasons it might fail, such as code already being at that address, or the constructor itself reverting

```solidity
        if (addr == address(0)) {
            revert Create2FailedDeployment();
        }
    }
```

Lastly, let us create a view function named `computeAddress`. This function should take in the `salt` and `bytecode` as parameters, and return the address of the contract that would be deployed using the `deploy` function:

```solidity

>  function computeAddress(bytes32 salt, bytes32 bytecodeHash) external view returns (address addr) {

Reviewer: ```suggestion
 function computeAddress(bytes32 salt, bytes32 creationCodeHash) external view returns (address addr) {
```

 }
```

Inside the function, paste the following code that uses in-line assembly to calculate the address by performing the same calculations as the CREATE2 opcode would:

```solidity
        address contractAddress = address(this);
        

>         assembly {

Reviewer: Why do we need assembly here? It will be more readable to just do `keccak256(abi.encodePacked(bytes1(0xff), address(this), salt, creationCodeHash))` (and since it's a `view` function typically used off-chain we don't need to worry about gas)

Author: Good point Matt. But pls pls let me keep this. I did this for two reasons:

1. The OZ implementation uses inline assembly, and I directly borrow from their library.
2.  Also I would really really like to be able to say I pushed some assembly code to the Foundry book :(

Please let me have this one

> ℹ️ **Note**
>
> Recall that all the params passed to the `computeAddress` function are 32 bytes long, and are stored in memory as 32 byte values. However
> an address in Solidity is 20 bytes long, and is stored in memory as a 32 byte value, where the first 12 bytes are replaced by 0s.
> Therefore, when we need to skip 12 bytes to point to the actual address.

5. `let start := add(ptr, 0x0b)` creates a new variable `start` that points to the memory location `ptr + 0x0b`, i.e. `ptr + 11 bytes`.
6. Lastly, the mstore8 opcode can be used to store a single byte at a memory location. Here, we are storing the value `0xff` at the memory location pointed to by `start`, which occupies the 12th byte of the memory slot.
7. With all the values packed into their correct memory locations, we can now call `keccak256` on the memory slot starting at `start`, and pass in the length of the memory slot as the second parameter. This will return a 32 byte hash, which we can truncate to get the final address.

> ℹ️ **Note**
>
> You can check out the complete code for this factory implementation [here](https://github.com/Genesis3800/CREATE2Factory/blob/main/src/Create2.sol).

> > Also check out OpenZeppelin's [CREATE2 library implementation](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/793d92a3331538d126033cbacb1ee5b8a7d95adc/contracts/utils/Create2.sol), which has been used as inspiration for this tutorial.

Reviewer: We should link to the forge-std address computation helpers also: https://github.com/foundry-rs/forge-std/blob/f73c73d2018eb6a111f35e4dae7b4f27401e9421/src/StdUtils.sol#L122-L134

Author: Will do

Author: @mds1 adding this exact Github link since this util function does not seem to be documented in the Foundry book.

> ### Fork testing

Reviewer: I'd just make this section use a local test instead of a fork test—there's nothing in particular about this tutorial that requires fork testing, and since we already have docs on fork testing, it'd be clearer to have a tight scope and keep this guide focused on create2 specifically

Author: Well then it would make sense to just reduce the test file to one function only. The two functions were meant to test CREATE2 deployments on each chain. If fork testing is not required, I'll cut it down.

You will need to blockchain nodes  in order to run fork tests.
[Alchemy](https://www.alchemy.com/) supports forking on their free tier, so we will use that for this tutorial.
Create a `.env` file at the root and fill out these values:

```bash
ALCHEMY_SEPOLIA_URL=''
ALCHEMY_MUMBAI_URL=''
PRIVATE_KEY=
```

Fill out these values, and run `source .env` to load them into your environment.

Create a file named `Create2.t.sol` Inside the `test` directory.
Initialize a contract named `Create2Test` like this:

```solidity
// SPDX-License-Identifier: UNLICENSED

> pragma solidity ^0.8.13;

Reviewer: Above we use MIT and `^0.8.20`—I'm not sure what other parts of the book use but let's be consistent with those

Reviewer 2: @Genesis3800 pls take a look at this one again, it is still the case. when this is fixed this lgtm

import {Test} from "forge-std/Test.sol";
import {Create2} from "../src/Create2.sol";

contract Create2Test is Test {

}
```

> Initialize the following state variables and the `setup()` function:

Reviewer: ```suggestion
Initialize the following state variables and the `setUp()` function:
```

        mumbaiFork = vm.createFork(ALCHEMY_MUMBAI_URL);
    }
```

The `createFork` cheatcode allows us to create a fork of the blockchain at a given URL.
We will activate these forks as required inside the individual test functions.

Create a new function named `testSepoliaDeploy()` that:

1. Activates the `sepoliaFork` fork using the `selectFork` cheatcode.
2. Deploys a new instance of the `Create2` contract.
3. Deals out a 100 ETH to the a specific address that we will use to impersonate as the caller for all subsequent calls using the `prank` cheatcode.
4. Sets up the `salt` and `bytecode` params
5. Uses the previously deployed instance of the `Create2` contract to deploy a new instance of the `Create2` contract at a deterministic address.
6. Checks if the contract was deployed at the correct address, by asserting that the computed address is equal to the deployed address.

```solidity
    function testSepoliaDeploy() public {

        vm.selectFork(sepoliaFork);
        assertEq(vm.activeFork(), sepoliaFork);

        create2 = new Create2();

        vm.deal(address(0x1), 100 ether);
        vm.startPrank(address(0x1));
        
        bytes32 salt = "12345";

>         bytes memory bytecode = abi.encodePacked(vm.getCode("Create2.sol:Create2"));

Reviewer: If you import the contract you can also get the creation code using `type(MyContract).creationCode` to avoid needing a cheat
